widget/static/js/widget.js

```diff
@@ -110,36 +150,216 @@ DiscoverUniWidget.prototype = {
+
+    generateLink: function() {
+        coursePageBase = '{{domain_name}}/course-details/{{uni_id}}/{{course_id}}/{{mode}}/';
+        coursePage = coursePageBase.replace('{{uni_id}}', this.institution);
+        coursePage = coursePage.replace('{{course_id}}', this.course);
+        coursePage = coursePage.replace('{{mode}}', this.kismode);
+        return coursePage;
+    }
+}
+
+var DataWidget = function(targetDiv, courseData, language, languageKey, kismode, generateLink) {
+    this.targetDiv = targetDiv;
+    this.courseData = courseData
+    this.language = language;
+    this.languageKey= languageKey;
+    this.kismode =  kismode;
+    this.generateLink = generateLink;
+    this.setup();
+}
+
+DataWidget.prototype = {
+    setup: function() {
+        this.slideIndex = 0;
+        this.renderDataLead();
+        this.renderCTABlock();
+    },
+
+    renderDataLead: function() {
         var leadNode = document.createElement('div');
         leadNode.classList.add('widget-lead');
+
+        leadNode.appendChild(this.renderSatisfactionSlide());
+        leadNode.appendChild(this.renderExplanationSlide());
+        leadNode.appendChild(this.renderWorkSlide());
+
+        this.targetDiv.appendChild(leadNode);
+        this.carousel();
+    },
+
+    renderSatisfactionSlide: function() {
+        var isNotAggregated = this.courseData.statistics.nss[0].aggregation_level === 14 ||
+                                this.courseData.statistics.nss[0].aggregation_level === 24
+
+        var slideNode = document.createElement('div');
+        slideNode.classList.add('lead-slide', 'fade');
+        slideNode.id = 'satisfaction';
+
+        var statNode = document.createElement('div');
+        statNode.classList.add('stat');
+
         var titleNode = document.createElement('h1');
         titleNode.classList.add('title');
+        var percentage = this.courseData.statistics.nss[0].question_27.agree_or_strongly_agree + '%';
+        var title = document.createTextNode(percentage);
+        titleNode.appendChild(title);
+
+        var introNode = document.createElement("p");
+        introNode.classList.add('intro');
+        var intro = document.createTextNode(CONTENT.satisfactionIntro[this.language]);
+        introNode.appendChild(intro);
+
+        statNode.appendChild(titleNode);
+        statNode.appendChild(introNode);
 
-        var percentage = courseData.statistics.nss[0].question_27.agree_or_strongly_agree + '%';
+        slideNode.appendChild(statNode);
+        slideNode.appendChild(this.renderCourseDetails(isNotAggregated));
+
+        return slideNode;
+    },
+
+    renderExplanationSlide: function() {
+        var isNotAggregated = this.courseData.statistics.nss[0].aggregation_level === 14 ||
+                                this.courseData.statistics.nss[0].aggregation_level === 24
+
+        var slideNode = document.createElement('div');
+        slideNode.classList.add('lead-slide', 'fade');
+        slideNode.id = 'satisfaction';
```

Comment:
Probably my lack of knowledge about JS, but should the slideNode.id still be 'satisfaction' i.e., the same as in renderSatisfactionSlide?

Reply:
Again, probably due to my lack of knowledge of how things can be done in browser JS, but there seems to be a lot of repeated code in the render... functions. Could this be pulled out into a separate function?

Reply:
Well spotted.
No it should not. Copy/paste fail.

Reply:
Have done a refactor to remove the duplicate code in to separate functions

Reply:
:+1: